### DISCUSSION TITLE
How do you make a keyBy on an object?

### DISCUSSION THREAD

**Discussion by LunaticMuch** [ramda/ramda#3377] [Q&A]
I am looking at migrating out from Lodash, but as expected in some cases it's more tricky.
The code I have is using a `keyBy` as

```
schema.types = _.keyBy(schema.types, "id");
```

Which could be swapped with

```
schema.types = R.indexBy(R.prop("id"), schema.types);
```

However, `schema.types` is an object, in a form like (when it's the input):

```
schema.types: {
  a: { id: "a::a", name: ""},
  b: { id: "b::b", name: ""},
  c: { id: "c::C", name: ""},
}
```
With the keyBy it gets out as
```
schema.types: {
  a::a: { id: "a::a", name: ""},
  b::b: { id: "b::b", name: ""},
  c::c: { id: "c::C", name: ""},
}
```

So natively, can't go. Which option could I use?

**Comment by adispring** [comment]
try `groupBy`

**Reply by LunaticMuch** [reply]
I did it already, but see

```
node_modules/ramda/es/internal/_createReduce.js:29
    throw new TypeError('reduce: list must be array or iterable');
          ^
TypeError: reduce: list must be array or iterable
```

**Comment by kedashoe** [comment]
Hello @LunaticMuch . You've written your input as
```
schema.types: {
  { id: "a", name: ""},
  { id: "b", name: ""},
  { id: "c", name: ""},
}
```
, but this isn't valid javascript (types is an object with values but no keys). Is the types field meant to be an array?
```
schema.types: [
  { id: "a", name: ""},
  { id: "b", name: ""},
  { id: "c", name: ""},
]
```
in which case what you've written, `R.indexBy(R.prop("id"), schema.types);`, looks like it will work?

**Comment by LunaticMuch** [comment]
My bad, I updated the example. It's not an array anyway. Each object has a key, the key needs to be replaced with the `id` property included in each object

**Reply by kedashoe** [reply]
Ah ok, so in this case, we can add a call to [R.values](https://ramdajs.com/docs/#values) in there
```
schema.types = R.indexBy(R.prop("id"), R.values(schema.types));
```
or formatted another way
```
schema.types = R.pipe(
  R.prop("types"),
  R.values,
  R.indexBy(R.prop("id"))
)(schema);
```
